### PR TITLE
blog post draft "Fixing a bug in PyPy's incremental GC"

### PR DESCRIPTION
**Pull request by cfbolz**

### CODE REVIEW

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+
+A few weeks ago, we got [two more](https://github.com/PyO3/pyo3/issues/3766)
+[bug reports](https://github.com/orgs/pypy/discussions/4923), the last one by
+the authors of the [nanobind](https://nanobind.readthedocs.io/) binding
+generator, with the same symptoms: crash in AST rewriting, only on CI. I
+decided to make a more serious push to try to find the bug this time.
+Ultimately the problem turned out to be several bugs in PyPy's garbage
+collector (GC) that had been there since its inception in
+[2013](https://www.pypy.org/posts/2013/10/incremental-garbage-collector-in-pypy-8956893523842234676.html)
+(started by [Andrew Chambers](https://acha.ninja/)). Understanding the
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:31]: It sounds a little like you are blaming Andrew. Maybe leave this part out?
```suggestion
Understanding the
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+Since the bug was so much work to find, I thought I'd write a blog post about
+it.
+
+The blog post consists of three parts: first a chronological description of
+what I did to find the bug, a technical explanation of what goes wrong, some
+reflections on the bug (and then a bonus bug I also found in the process).
+
+# Finding the Bug
+
+I was starting from the failing [nanobind
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:43]:
```suggestion
I started from the failing [nanobind
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+CI](https://github.com/wjakob/nanobind/actions/runs/8234561874/job/22516568891)
+runs that ended with a segfault of the PyPy interpreter. This was only an
+intermittent problem, not every run was failing. When I tried to just run the
+test suite locally, I couldn't get it to fail. Therefore at first I tried to
+learn more about what was happening by looking on the CI runners.
+
+## Running on CI
+
+I forked the nanobind repo and hacked the CI script in order to get it to use a
+PyPy build with full debug information and more assertions turned on. In order
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:53]: Is there a link how to do this?
```suggestion
PyPy build with [full debug information and more assertions turned on](https://doc.pypy.org/en/latest/build.html#making-a-debug-build-of-pypy). In order
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+I experimented with hacking our faulthandler implementation to also give me a
+C-level callstack, but that didn't work as well as I hoped.
+
+Then I tried to run [gdb](https://sourceware.org/gdb/) on CI to try to get it
+to print a C callstack at the crash point. You can get gdb to execute commands
+as if typed at the prompt with the `-ex` commandline option, I used something
+like this:
+
+```
+gdb -ex "set confirm off" -ex "set pagination off" -ex "set debuginfod enabled off" -ex run -ex where -ex quit --args <command> <arguments>
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:77]: Will this render more nicely if it is wrapped?
```suggestion
gdb -ex "set confirm off" -ex "set pagination off" -ex \
    "set debuginfod enabled off" -ex run -ex where -ex quit \
    --args <command> <arguments>
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+write some helper commands to understand the state of the GC heap (rr is an
+extension of GDB, so the GDB scripting API works in rr too).
+
+The first (small) helper command I wrote with the GDB scripting API was a way
+to pretty-print the currently active GC flags of a random PyPy object, starting
+just from the pointer. The more complex command I wrote was an object tracer,
+which follows pointers to GC objects starting from a root object to explore the
+object graph. The object tracer isn't complete, it doesn't deal with all the
+complexities of PyPy's GC. But it was good enough to help me with my problem, I
+found out that the corrupted object was stored in an array.
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:188]: Is there a way to preserve the helpers in a wiki or as part of the pypy repo?

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+		self.gc.collect_step()
+	# used to crash, node got collected
+	assert target[0].x == 5
+
+```
+
+One of the fun properties of testing our GC that way is that all the memory is
+emulated. The crash in the last line of the test isn't a segfault at all,
+instead you get a nice exception saying that you tried to access a freed chunk
+of memory and you can then debug this with a python2 debugger.
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:254]: I like the way you (subtly?) (not so subtly?) drive home the point that writing in RPython has advantages for dense complicated code like GC implementations.

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+thread).
+
+The incremental GC uses [tri-color
+marking](https://en.wikipedia.org/wiki/Tracing_garbage_collection#Tri-color_marking)
+to reason about the reachable part of the heap during the marking phase, where
+old objects can be:
+
+- black: already marked, reachable, definitely survives the collection
+- grey: will survive, but still needs to be marked
+- white: potentially dead The color of every object is encoded by setting flags
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:343]:
```suggestion
- white: potentially dead 

The color of every object is encoded by setting flags
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+gray. During a major collection cycle, the GC will trace gray objects, until
+none are left. To trace a gray object, all the objects it references have to be
+marked grey if they are white so far. After a grey object is traced, it can be
+marked black (because all the referenced objects are now either black or gray).
+Eventually, there are no gray objects left. At that point (because no white
+object can be reached from a black one) all the white objects are known to be
+unreachable and can therefore be freed.
+
+The GC is incremental because every collection step will only trace a limited
+amount of gray object, before giving control back to the program. This leads to
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:357]:
```suggestion
number of gray objects, before giving control back to the program. This leads to
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+a problem: if an already traced (black) object is changed between to marking
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:358]:
```suggestion
a problem: if an already traced (black) object is changed between two marking
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+steps of the GC, the program can mutate that object and write a new reference
+into one of its field. This could lead to an invariant violation, if the
+referenced object is white. Therefore, the GC uses the write barrier (which it
+needs anyway to find references from old to young objects) to mark all black
+objects that are written into gray, and then trace them again at one of the
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:363]: Not clear what "written" is: "traced" or "modified".
```suggestion
needs anyway to find references from old to young objects) to mark all black
objects that are written into gray, and then trace them again at one of the
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 11]
```diff
@@ -0,0 +1,478 @@
...
+later collection steps.
+
+## The special write barrier of memcopy
+
+Arrays use a different kind of write barrier than normal objects. Since they
+can be arbitrarily large, tracing them can take a long time. Therefore it's
+potentially wasteful to trace them fully at a minor collection. To fix this.
+the array write barrier keeps more granular information about which parts of
+the array have been written to since the last collection step. Then only the
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:372]:
```suggestion
the array have been modified since the last collection step. Then only the
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 12]
```diff
@@ -0,0 +1,478 @@
...
+property-based test for the GC would have been a good idea given the complexity
+of the GC, and definitely something we did in other parts of our code at the
+time (just using the `random` module mostly, we started using hypothesis
+later).
+
+It's a bit of a mystery to me why this bug managed to be undetected for so
+long. Memcopy happens in a lot of pretty core operations of e.g. lists in
+Python (`list.extend`, to name just one example). To speculate, I would suspect
+that all the other preconditions for the bug occurring made it pretty rare:
+- the contend of an old list that is not yet marked needs to be copied into
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:413]:
```suggestion
- the content of an old list that is not yet marked needs to be copied into
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+- the source of the copy needs to also store an object that has no other
+  references 
+- the source of the copy then needs to be overwritten with other data 
+- then the next collection steps need to be happening at the right points
+- ...
+
+Given the complexity of the GC logic I also wonder whether some lightweight
+formal methods would have been a good idea. Formalizing some of the core
+invariants in B or TLA+ and then model checking them up to some number of
+objects would have found this problem pretty soon. There are also correctness
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:424]:
```suggestion
objects would have found this problem pretty quickly. There are also correctness
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+old-to-young pointers) to notice when a pre-built object gets modified for the
+very first time. If that happens, it gets added to the set of pre-built objects
+that gets counted as a root, and is therefore traced as a root at collections
+from then on.
+
+## The wrong assertion
+
+The assertion that triggered when I turned on the GC debug mode was saying that
+the GC found a reference from a black to a white object, violating its
+invariant. pre-built objects count as always black, as long as they have never
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:459]:
```suggestion
invariant. Pre-built objects count as always black, as long as they have never
```

File: posts/2024/03/fixing-bug-incremental-gc.md (hunk 4 above)
Comment (cfbolz) [reply posts/2024/03/fixing-bug-incremental-gc.md:77]: yeah, it should, I'll try it.

File: posts/2024/03/fixing-bug-incremental-gc.md (hunk 5 above)
Comment (cfbolz) [reply posts/2024/03/fixing-bug-incremental-gc.md:188]: good question! I'll look into it

File: posts/2024/03/fixing-bug-incremental-gc.md (hunk 6 above)
Comment (cfbolz) [reply posts/2024/03/fixing-bug-incremental-gc.md:254]: I think it is actually an advantage, yes. the fact that we have debugged a bunch of GC problems in GDB is kind of a sign that we under-tested our rpython code as python2 code.

File: posts/2024/03/fixing-bug-incremental-gc.md (hunk 14 above)
Comment (cfbolz) [reply posts/2024/03/fixing-bug-incremental-gc.md:459]: wow, this sentence is pretty convoluted, I think I'll try to reformulate it

File: posts/2024/03/fixing-bug-incremental-gc.md (hunk 10 above)
Comment (cfbolz) [reply posts/2024/03/fixing-bug-incremental-gc.md:367]: I replaved "written into" with "modified" here.

File: posts/2024/03/fixing-bug-incremental-gc.md (hunk 14 above)
Comment (cfbolz) [reply posts/2024/03/fixing-bug-incremental-gc.md]: done

**Comment by mattip** [comment]
Cool. I like the animations.

File: posts/2024/03/fixing-bug-incremental-gc.md (hunk 5 above)
Comment (cfbolz) [reply posts/2024/03/fixing-bug-incremental-gc.md:200]: I've added a gist link to the full code

**Comment by mattip** [comment]
Final version visible [here](https://deploy-preview-112--keen-mestorf-442210.netlify.app/posts/2024/03/fixing-bug-incremental-gc), looking through one last time for obvious errors and omissions

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 15]
```diff
@@ -0,0 +1,569 @@
...
+it.
+
+The blog post consists of three parts: first a chronological description of
+what I did to find the bug, a technical explanation of what goes wrong, some
+reflections on the bug (and then a bonus bug I also found in the process).
+
+# Finding the Bug
+
+I started from the failing [nanobind CI
+runs](https://github.com/wjakob/nanobind/actions/runs/8234561874/job/22516568891)
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:44]: This link will disappear in a few months...

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 16]
```diff
@@ -0,0 +1,569 @@
...
+confirmed that at the point of crash no C-extension had been imported yet.
+
+## Using `rr`
+
+I still couldn't get the bug to happen in GDB, so the tool I tried next was
+[rr](https://rr-project.org/). rr can record the execution of a program and
+later replay it arbitrarily often. This gives you a time-traveling debugger
+that allows you to execute the program backwards in addition to forwards.
+Eventually I managed to get the crash to happen when running the tests with `rr
+record --chaos` (`--chaos` randomizes some decisions that rr takes, to try to
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:150]: Line break in the code fragment renders as a line break, so avoid that
```suggestion
Eventually I managed to get the crash to happen when running the tests with
`rr record --chaos` (`--chaos` randomizes some decisions that rr takes, to try to
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 17]
```diff
@@ -0,0 +1,569 @@
...
+Next source of bugs to exclude where C-extensions. Since those were the tests
+of nanobind, a framework for creating C-extension modules I was a bit worried
+that the bug might be in our emulation of CPython's C-API. But running PyPy
+with the `-v` option (which will print all the imports as they happen)
+confirmed that at the point of crash no C-extension had been imported yet.
+
+## Using `rr`
+
+I still couldn't get the bug to happen in GDB, so the tool I tried next was
+[rr](https://rr-project.org/). rr can record the execution of a program and
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:146]: the link gets a bit lost, maybe something like this? Or am I getting too carried away :)
```suggestion
[rr, the "reverse debugger"](https://rr-project.org/). rr can record the execution of a program and
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 18]
```diff
@@ -0,0 +1,569 @@
...
+logic itself was wrong. I have fixed that in the meantime too, the details
+of that are in the bonus section at the end of the post.
+
+## Using GDB scripting to find the real bug
+
+After that disaster I went back to the earlier rr recording without GC assertions
+and tried to understand in more detail why the GC decided to free an object
+that was still being referenced. To be able to do that I used the [GDB Python
+scripting
+API](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Python-API.html) to
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:190]: This link renders well, so the problem above is not connected to having an `.html` suffix.

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 19]
```diff
@@ -0,0 +1,569 @@
...
+problems with memcopying an array.
+
+In addition, the test also found corner cases in my fix. There was a situation
+that I hadn't accounted for, which the test found after eventually.
+I also plan on adding a bunch of other GC features as steps in the
+test to stress them too (for example weakrefs, identity hashes, pinning, maybe
+finalization).
+
+At the point of publishing this post, the fixes got merged to the 2.7/3.9/3.10
+branches of PyPy, and we'll have to do a release with the fixes at some point.
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:365]:
```suggestion
branches of PyPy, and will be part of the next release (v7.3.16).
```

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 20]
```diff
@@ -0,0 +1,569 @@
...
+```python
+def arraycopy(source, dest, source_start, dest_start, length):
+    gc_writebarrier_before_array_copy(source, dest, source_start, dest_start, length)
+    raw_memcopy(cast_to_voidp(source) + source_start,
+                cast_to_voidp(dest) + dest_start,
+                sizeof(itemtype(source)) * length)
+```
+
+(this is really a rough sketch. The [real
+code](ttps://github.com/pypy/pypy/blob/789f964fff59c722b0872abcdc56d2b1373a9f3b/rpython/rlib/rgc.py#L365)
```
Comment (mattip) [inline posts/2024/03/fixing-bug-incremental-gc.md:462]:
```suggestion
code](https://github.com/pypy/pypy/blob/789f964fff59c722b0872abcdc56d2b1373a9f3b/rpython/rlib/rgc.py#L365)
```

**Comment by mattip** [comment]
thanks, just a few last tweaks and one unsolved mystery in rendering a link. I really like the complete sleuthing including the dead ends and false flags. It adds so much to the story.

File: posts/2024/03/fixing-bug-incremental-gc.md [hunk 21]
```diff
@@ -0,0 +1,569 @@
...
+nowhere.
+
+A few weeks ago, we got [two more](https://github.com/PyO3/pyo3/issues/3766)
+[bug reports](https://github.com/orgs/pypy/discussions/4923), the last one by
+the authors of the [nanobind](https://nanobind.readthedocs.io/) binding
+generator, with the same symptoms: crash in AST rewriting, only on CI. I
+decided to make a more serious push to try to find the bug this time.
+Ultimately the problem turned out to be several bugs in PyPy's garbage
+collector (GC) that had been there since its inception in
+[2013](https://www.pypy.org/posts/2013/10/incremental-garbage-collector-in-pypy-8956893523842234676.html).
```
[the remark this answers is not in the document]
Comment (cfbolz) [reply posts/2024/03/fixing-bug-incremental-gc.md:30]: I wonder whether nikola is configured to do something special for links that go to the same domain. locally at least replacing it with a relative link fixes it, will try to push

File: posts/2024/03/fixing-bug-incremental-gc.md (hunk 15 above)
Comment (cfbolz) [reply posts/2024/03/fixing-bug-incremental-gc.md:44]: I think I don't care too much. we link to the discussion too, that's good enough when the action run gets removed

**Comment by cfbolz** [comment]
> thanks, just a few last tweaks and one unsolved mystery in rendering a link. I really like the complete sleuthing including the dead ends and false flags. It adds so much to the story.

thanks a lot :blush:
